data-pipeline/telemetry-location-updater/src/main/java/org/ekstep/ep/samza/service/TelemetryLocationUpdaterService.java

```diff
@@ -10,42 +11,128 @@
 import org.ekstep.ep.samza.task.TelemetryLocationUpdaterSink;
 import org.ekstep.ep.samza.task.TelemetryLocationUpdaterSource;
 import org.ekstep.ep.samza.util.DeviceProfileCache;
+import org.ekstep.ep.samza.util.Path;
+import org.ekstep.ep.samza.util.RedisConnect;
+import redis.clients.jedis.Jedis;
+import com.google.gson.Gson;
+import redis.clients.jedis.exceptions.JedisException;
+
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
 
 public class TelemetryLocationUpdaterService {
 
 	private static Logger LOGGER = new Logger(TelemetryLocationUpdaterService.class);
 	private DeviceProfileCache deviceProfileCache;
+	private int userStoreDb;
+	private RedisConnect redisConnect;
+	private Jedis userDataStoreConnection;
 	private JobMetrics metrics;
+	private Gson gson = new Gson();
+	private Path path = new Path();
+	private Type mapType = new TypeToken<Map<String, Object>>() { }.getType();
 
-
-	public TelemetryLocationUpdaterService(DeviceProfileCache deviceLocationCache, JobMetrics metrics) {
+	public TelemetryLocationUpdaterService(DeviceProfileCache deviceLocationCache, JobMetrics metrics, RedisConnect redisConnect, Config config) {
 		this.deviceProfileCache = deviceLocationCache;
 		this.metrics = metrics;
+		this.userStoreDb = config.getInt("redis.database.userStore.id", 4);
+		this.redisConnect = redisConnect;
+		this.userDataStoreConnection = this.redisConnect.getConnection(userStoreDb);
 	}
 
 	public void process(TelemetryLocationUpdaterSource source, TelemetryLocationUpdaterSink sink) {
 		try {
 			Event event = source.getEvent();
-			// Add device location details to the event
-			updateEventWithIPLocation(event);
-			sink.toSuccessTopic(event);
+			String did = event.did();
+			DeviceProfile deviceProfile = null;
+			if (did != null && !did.isEmpty()) {
+				// check for user profile
+				Map<String, String> derivedLocation = getLocationFromUserCache(event);
+				// get device profile from cache
+				deviceProfile = deviceProfileCache.getDeviceProfileForDeviceId(did);
+
+				// get user declared location if user profile is empty
+				if (derivedLocation.isEmpty()) {
+					derivedLocation = getUserDeclaredLocation(deviceProfile);
+				}
+				// get ip reso lved location if user declared location is empty
+				if (derivedLocation.isEmpty()) {
+					derivedLocation = getIpResolvedLocation(deviceProfile);
+				}
+				// Add derived location to telemetry
+				event.addDerivedLocation(derivedLocation);
```

Comment:
There is a possibility that the location at this point can still be empty. Can we have a push out metrics such as - totalEvents, userCacheHit, userDeclaredHit, ipLocationHit and noCacheHit?